storage/merkle/proof.go

```diff
@@ -0,0 +1,133 @@
+package merkle
+
+import (
+	"bytes"
+
+	"github.com/onflow/flow-go/ledger/common/bitutils"
+)
+
+// Proof captures all data needed for proving inclusion of a single value inserted under key `Key` into the merkle trie
+type Proof struct {
+	// Key used to insert and look up the value
+	Key []byte
+	// Value stored in the trie for the given key
+	Value []byte
+	// IsAShortNode holds bits of data to determine short nodes versus full nodes while tranversing the
+	// trie downward. if the bit is set to 1, it means that we have reached to a short node, and
+	// if is set to 0 means we have reached a full node.
+	IsAShortNode []byte
+	// SkipBits is read when we reach a short node, and the value represents number of bits that were skipped
+	// by the short node (shortNode.count)
+	SkipBits []uint16
+	// InterimHashes is a slice of hash values, every value is read when we reach a full node (hash value of the siblings)
+	InterimHashes [][]byte
+}
+
+// Verify verifies the proof by constructing the hash values bottom up and cross check
+// the constructed root hash with the given one.
+// if the proof is valid it returns true and false otherwise
+func (p *Proof) Verify(expectedRootHash []byte) (bool, error) {
+
```

Comment:
I would suggest to add detailed documentation here what conditions are checked and where. It took me literally half an hour to convince myself that these checks are sufficient (except for an edge case, which I'll comment on below).

The way I would document the security conditions is as follows
```suggestion

	// A valid proof as to satisfy the following consistency conditions:
	// 1. A valid inclusion proof represents a full path through the merkle tree.
	//    We separate the path into a sequence of interim vertices and a tailing leaf.
	//    For interim vertex (with index i, counted from the root node) along the path,
	//    the proof as to contain the following information:
	//      (i) whether the vertex is a short node (InterimNodeTypes[i] == 1) or
	//          a full node (InterimNodeTypes[i] == 0)
	//     (ii) for each short node, we need the number of bits in the node's key segment
	//          (entry in ShortPathLengths)
	//    (iii) for a full node, we need the hash of the sibling that is _not_ on the path
	//          (entry in SiblingHashes)
	//    Hence, len(ShortPathLengths) + len(SiblingHashes) specifies how many _interim_
	//    vertices are on the merkle path. Consequently, we require the same number of _bits_
	//    in InterimNodeTypes. Therefore, we know that InterimNodeTypes should have a length
	//    of `(numberBits+7)>>3` _bytes_.
	// 2. The key length (measured in bytes) has to be in the interval [1, 8192].
	//    Furthermore, each interim vertex on the merkle path represents:
	//    * either a single bit in case of a full node:
	//      we expect InterimNodeTypes[i] == 0
	//    * a positive number of bits in case of a short node:
	//      we expect InterimNodeTypes[i] == 1 
	//      and the number of bits is encoded in the respective element of ShortPathLengths
	//    Hence, the total key length _in bits_ should be: len(SiblingHashes) + sum(ShortPathLengths)
```

Reply:
added